### ISSUE TITLE
Rule:ClassNamingConventions - class suffixed Util, Utils & Helper is known bad practice

### ISSUE THREAD

**Issue by derekm** [pmd/pmd#1710]
PMD is enforcing bad practices. PMD should not enforce bad practices.

**Comment by derekm** [comment]
Citation needed: https://twitter.com/kevlinhenney/status/715093313301454849

**Comment by rlangenfeld** [comment]
Agreed, in general I think it's time to evaluate some of the things that PMD considers to be "best practice" like this and bring them up to modern day standards or remove them all together.

**Comment by jsotuyod** [comment]
@darekm thanks for the report.

However, I don't think there is anything for us to do here.

Firstly, what one could argue is a bad practice is not the naming itself, but rather having classes that are not instantiable (which is a contradiction to OOP principles). Your "source" is quite lacking on arguments for his opinion on names themselves.

Utility and helper classes are actually well known concepts that are well documented:

https://en.m.wikipedia.org/wiki/Helper_class

Moreover, PMD enforces nothing. PMD is a tool to help developers enforce conventions and rules they like. It's up to the developer to set up the ruleset according to these preferences, and choose which rules, and which with configuration (ie: the rule allows to configure what names should apply to utility classes). The fact that PMD ships this rule with this configuration by default (again, enabling the rule is up to the user) is simply a sign that calling utility classes for what they are is a common enough practice to make sense as a default.

If you don't personally share or like such a rule, PMD is flexible enough for you to configure your ruleset accordingly. Shall you need any help in doing so, feel free to ask or check our docs.

**Comment by derekm** [comment]
@jsotuyod -- We have simply `<rule ref="rulesets/java/naming.xml/ClassNamingConventions"/>` in our ruleset, and this fascist/totalitarian rule is now breaking our builds unnecessarily and unjustly.

This new helper class detection and name enforcement rule should be removed in its entirety.

**Comment by ivangreene** [comment]
IMO PMD should just enforce the established Java conventions of asserting that utility classes end in 's' (Objects, Collections, Arrays, ...)

**Comment by derekm** [comment]
@jsotuyod -- Maybe you can set the default pattern for `utilityClassPattern` to `[A-Z][a-zA-Z0-9]*` like all the other properties of `ClassNamingConventions`.

`utilityClassPattern` shouldn't have `Util`, `Utils` or `Helper` suffix requirement for the same reasons `interfacePattern` doesn't have an `I` prefix, it is obsolete practice and a code smell.

This way you allow your users to opt into fascism instead of having it forced upon them.

**Comment by oowekyala** [comment]
@derekm  I don't think it's a code smell to use a well known naming convention to make explicit that a class is a static helper class, but I understand where you're coming from. 

This part of the rule has been getting a lot of hate just because nobody agrees about what's a utility class, or how they should be named. Ultimately the rule is made to be configured by the user, so I'm inclined to go with the suggestion to not enforce any specific pattern, document what the rule considers a utility class, and stop dealing with those overly opinionated reports. @jsotuyod Wdyt?

**Comment by jsotuyod** [comment]
@derekm first of, I'd like to gently ask you to review the words you select more carefully. Fascism / totalitarianism have little to do with Open Source Software. There are plenty of people suffering under such regimes in the world, and comparing this situation to theirs is a disrespectful and inconsiderate trivialization of their condition. Also, being nice, constructive and respectful to the work of others goes a long way when wanting them to cooperate with you, specially when it's for free and for no other incentive than love for what they do. 

> This new helper class detection and name enforcement rule should be removed in its entirety.

We have revised rule defaults on several occasions, even this very same property of this rule had it's default relaxed some time ago after suggestions from other users (who by the way, agreed with the rule, but simply found it too restrictive in allowed options).

Removing rules or functionality on the other hand, is something we don't do lightly. We actually only do so when there is clear evidence the rule is actually harmful **for everyone**… but that's hardly ever the case for a *codestyle* rule, as any code style is by default opinionated (even the default Java naming conventions such us using lowercase for package names, or camelcase for variable names is opinionated, even if recommended by Oracle / Sun). Removing functionality for the sake of an opinionated request is something we will **never** do. Just the same way, any user that actually likes this rule could come and **demand** the exact opposite with just as valid arguments as your own: *I don't like it and it shouldn't be this way!* 😠

> IMO PMD should just enforce the established Java conventions of asserting that utility classes end in 's' (Objects, Collections, Arrays, ...)

This has come up on the recent past… actually this *"established"* convention, was a rarity not too long ago. Up until Java 7, there were little uses for this aside from Java's `Collections` and some Guava classes. One could argue that using a `Utils` suffix was by far more common, citing hugely popular libraries such as Spring Framework or Apache Commons that to this day still follow such convention; so enforcing the `s` over the other doesn't seem less arbitrary.

Non the less, even allowing the `s` along the other options had the shortcoming of false negatives, as we don't really care about the name ending in `s`, but rather that the class name is a plural (a class named `Mass` would comply, but it's a singular name, we would expect `Masses`)… detecting plurals has a wide variety of problems due to irregular forms and was deemed not worth it at the time; but of course this is something we may revise once again; specially if there is a neat suggestion around this problem!

@oowekyala I think any default is opinionated, and the way around it remains the same: customize the rule to your liking. I am torn about changing the default value, as it may yield less opinionated reports, but not because of more satisfied users, but simply because of the same reason we get these reports in the first place: people not investing their time in actually customizing their rulesets to their liking. If we change the default, they will not see the utility classes reported, will never consider these as "false negatives", and will not benefit from it… Having people customize their rulesets and doing it as easily as possible remains the main problem. A "noisy" default may be more effective in achieving that than a silent one; but I agree this is certainly the least desirable way to achieve it. Non the less, if other maintainers agree with it, I'm ok with the change overall; just hope our users currently relying on this behavior are diligent in checking the PMD changelog and updating their rulesets… otherwise we are in an egg and chicken situation were we dumb down rules for everyone and just power users get the best out of it.


@derekm also of note, this *"new"* helper class detection and name enforcement as you call it, has been in PMD for almost a year now (merged to master on April 8th, 2018), and been part of no less than 10 PMD releases; and each change to the rule has been properly documented in each release's changelog. I strongly suggest moving forward you try to keep up with PMD so you can make the most out of it! Each release tends to have several substantial changes and improvements. Plus, it's super easy to stay up to date, we do monthly releases and announce on all channels, from mailing lists to Twitter!

**Comment by derekm** [comment]
Well, I would have opened this ticket a year ago if we or the maven plugin had upgraded sooner!

Seriously though, nobody wants Util, Utils, Utility, Utilities, Helper or Helpers as a suffix in their class names.

Probably we will just have to abandon PMD, it forces us to litter our code with magical incantations to unbreak the build because it generally doesn't understand modern Java practices with 8+ and CDI, etc.

This suffix enforcement is a very dated concept.

Wikipedia should not be your source for digging in your heels on this one.

**Comment by adangel** [comment]
Hi @derekm ,

I want to refer you once more to the documentation - since PMD does not force you to use the rule:
* [Best Practices](https://pmd.github.io/pmd-6.12.0/pmd_userdocs_best_practices.html): essential take away from there:  "Choose the rules that are right for you"
* [Making (custom) rulesets](https://pmd.github.io/pmd-6.12.0/pmd_userdocs_making_rulesets.html)
* [Configuring rules](https://pmd.github.io/pmd-6.12.0/pmd_userdocs_configuring_rules.html)
* [Using Rule Sets with maven-pmd-plugin](http://maven.apache.org/plugins/maven-pmd-plugin/examples/usingRuleSets.html)

In your case, you need to configure the properties of the rule ClassNamingConventions, since you don't want to use the defaults. This could look like so:

```xml
<rule ref="category/java/codestyle.xml/ClassNamingConventions">
    <properties>
        <property name="utilityClassPattern" value="[A-Z][a-zA-Z0-9]*" />
    </properties>
</rule>
```

(That's what @jsotuyod already suggested)

If you want to do something constructive, you could suggest a documentation improvement for the [rule documentation of ClassNamingConvention](https://pmd.github.io/pmd-6.12.0/pmd_rules_java_codestyle.html#classnamingconventions). We could add a sentence, to make it clear, that not everybody agrees to these conventions (especially the utility class naming) and if you disagree, you need to define your own conventions via the properties.

Best regards,
Andreas
